services/thumbnails/pkg/thumbnail/storage/filesystem.go

```diff
@@ -62,14 +62,29 @@ func (s FileSystem) Put(key string, img []byte) error {
 	}
 
 	if _, err := os.Stat(imgPath); os.IsNotExist(err) {
-		f, err := os.Create(imgPath)
+		f, err := os.CreateTemp(dir, "tmpthumb")
 		if err != nil {
-			return errors.Wrapf(err, "could not create file \"%s\"", key)
+			return errors.Wrapf(err, "could not create temporary file for \"%s\"", key)
 		}
-		defer f.Close()
 
-		if _, err = f.Write(img); err != nil {
-			return errors.Wrapf(err, "could not write to file \"%s\"", key)
+		_, writeErr := f.Write(img) // write the thumbnail in the temporary file
+		f.Close()                   // close the file regardless of the error
+
+		// if there was a problem writing, remove the temporary file
+		if writeErr != nil {
+			if remErr := os.Remove(f.Name()); remErr != nil {
+				return errors.Wrapf(remErr, "could not cleanup temporary file for \"%s\"", key)
+			}
+			return errors.Wrapf(writeErr, "could not write to temporary file for \"%s\"", key)
+		}
+
```

Comment:
AFAIK, to be really on the save side here we should do a `f.Sync()` to make sure the data is actually written to disk.

Nit: Also after the rename block there should be a fsync call on the parent directory. (See e.g. https://lwn.net/Articles/457667/)

Reply:
> I might have some wrong assumptions... The `Close` method should flush the buffers, at least at user-space level, so after that it should be up to the kernel to decide when the data is written. It's true that something could happen and the data in the kernel-space might not be written (maybe an unexpected power failure). Even if the write is delayed by the kernel, I assume the rename still happens after write. Basically, I think the operations should be queued in the kernel.

(Disclaimer: This is where my Linux filesystem knowledge get's blurry)

The rename is just an operation on the file's parent directory. And mostly independent from the write. I am sure if they are guaranteed be ordered in the same sequence by the kernel. It seems (depending on a lot of ifs and whens) that under some circumstances you can endup with and empty dest file if you don't fsync before the rename. (I guess we're getting into hair-splitting mode here and I don't know if we need to protect the thumbnails service against this scenario, but if we want to be on the safe side I think we need an fsync. Also considering the fact that the storage might be living on a shared filesystem use by multiple thumbnails instances.)

> If both assumptions are right, the code should be fine. Even if the operations are delayed in the kernel, the thumbnail service will either send an error (missing file) or a complete thumbnail. Getting an "old" thumbnail (if the kernel delays too much) shouldn't be a problem.

Getting an "old" thumbnail shouldn't be an issue at all. The filenames a based on the checksum of the file. If a new thumbnail needs to be generated for a file, the checksum of the file will have changed and it gets written to a new file.

> There could be an issue where we could serve an "old" thumbnail if the server crashes while generating the new thumbnail for the updated file, but after a crash I don't think it's a big deal, and reuploading the file should regenerate the thumbnail so the user can easily fix it.
> 
> Based on this, I don't think we need to be so strict with the `Sync` call and let the kernel do what it needs to do.